packages/api/src/controllers/animalBatchController.js

```diff
@@ -0,0 +1,150 @@
+      try {
+        const { farm_id } = req.headers;
+        const rows = await AnimalBatchModel.query()
+          .where({ farm_id })
+          .whereNotDeleted()
+          .withGraphFetched('animal_batch_sex_detail');
+        return res.status(200).send(rows);
+      } catch (error) {
+        console.error(error);
+        return res.status(500).json({
+          error,
+        });
+      }
+    };
+  },
+
+  addAnimalBatches() {
+    return async (req, res) => {
+      const trx = await transaction.start(Model.knex());
+
+      try {
+        const { farm_id } = req.headers;
+        const result = [];
+
+        if (!Array.isArray(req.body)) {
+          await trx.rollback();
+          return res.status(400).send('Request body should be an array');
+        }
+
+        for (const animalBatch of req.body) {
+          if (animalBatch.custom_type_id) {
+            const customType = await CustomAnimalTypeModel.query().findById(
+              animalBatch.custom_type_id,
+            );
+            if (customType && customType.farm_id !== farm_id) {
+              await trx.rollback();
+              return res.status(403).send('Forbidden custom type does not belong to this farm');
+            }
+          }
+
+          if (animalBatch.default_breed_id && animalBatch.default_type_id) {
+            const defaultBreed = await DefaultAnimalBreedModel.query().findById(
+              animalBatch.default_breed_id,
+            );
+
+            if (defaultBreed && defaultBreed.default_type_id !== animalBatch.default_type_id) {
+              await trx.rollback();
+              return res.status(400).send('Breed does not match type');
+            }
+          }
+
+          if (animalBatch.custom_breed_id) {
+            const customBreed = await CustomAnimalBreedModel.query()
+              .whereNotDeleted()
+              .findById(animalBatch.custom_breed_id);
+
+            if (customBreed && customBreed.farm_id !== farm_id) {
+              await trx.rollback();
+              return res.status(400).send('Forbidden custom breed does not belong to this farm');
+            }
+
+            if (
+              customBreed.default_type_id &&
+              customBreed.default_type_id !== animalBatch.default_type_id
+            ) {
+              await trx.rollback();
+              return res.status(400).send('Breed does not match type');
+            }
+
+            if (
+              customBreed.custom_type_id &&
+              customBreed.custom_type_id !== animalBatch.custom_type_id
+            ) {
+              await trx.rollback();
+              return res.status(400).send('Breed does not match type');
+            }
+          }
+
+          if (animalBatch.count < 2) {
+            await trx.rollback();
+            return res.status(400).send('Batch count must be greater than 1');
+          }
+
+          if (animalBatch.animal_batch_sex_detail) {
+            let sexCount = 0;
+            const sexIds = [];
```

Comment:
I do that later with `arrayIsUnique()`... but not sure it would catch duplicates because its an object. 

I might need more help to learn about Sets!

Reply:
@SayakaOno and @Duncan-Brain wait I also would like to learn how to start using Sets! If I want to trim the duplicates off and don't need to return an error I can just convert to Set? I think I should be doing this is the group controller then...

```js
const uniqueAnimalIds = new Set(related_animal_ids);
const uniqueBatchIds = new Set(related_batch_ids);
```

Reply:
I am not 100% sure but if you do anything with those you may need to convert them to an array type vs a set type. I think its like Date() where theres some conversions needed

Reply:
I just meant that I would create a set instead of the `sexIds` array. (the array is just for making a set, correct?)
The current code:
```
const sexIds = [];
animalBatches.forEach(({ sex_id }) => sexIds.push(sex_id));
const isUnique = sexIds.length === new Set(sexIds).size;
```
I would create a set and add an id:
```
const sexIdsSet = new Set();
animalBatches.forEach(({ sex_id }) => sexIdsSet.add(sex_id));
const isUnique = animalBatches.length === sexIdsSet.size;
```
We could reduce the process of converting an array to set.

Reply:
@SayakaOno updated the check and marking as resolved !